### ISSUE TITLE
Update AbstractChecks to log DetailAST - JavadocMethod

### ISSUE THREAD

**Issue by rnveach** [checkstyle/checkstyle#7739]
Child of https://github.com/checkstyle/checkstyle/issues/5777

This issue specifically focuses on JavadocMethod.

**Comment by romani** [comment]
please be aware of https://github.com/checkstyle/checkstyle/issues/7740#issuecomment-599061545 , as most of violation are reported inside javadoc comment, we are not ready to support them in xpath. We should update only places where violation is logged on java AST node.

**Comment by wltan** [comment]
I am on it.

There are **8** uses of `log` within the check, numbered in order of appearance in the code.

## 1, 3, 6, 7
https://github.com/checkstyle/checkstyle/blob/e5c9f2c828c1098000def9a53ba0a28d14f103b6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L488-L490
https://github.com/checkstyle/checkstyle/blob/e5c9f2c828c1098000def9a53ba0a28d14f103b6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L790-L794
https://github.com/checkstyle/checkstyle/blob/e5c9f2c828c1098000def9a53ba0a28d14f103b6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L878-L880
https://github.com/checkstyle/checkstyle/blob/e5c9f2c828c1098000def9a53ba0a28d14f103b6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L890

This is logging the tags inside the javadoc block comment. This cannot be reached using DetailAST.

For item (7), the `lineNo` is referring to `@param lineNo the line number of the expected tag`, which is also within the block comment itself, and thus is not loggable.

## 2, 4, 5
https://github.com/checkstyle/checkstyle/blob/e5c9f2c828c1098000def9a53ba0a28d14f103b6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L509
https://github.com/checkstyle/checkstyle/blob/e5c9f2c828c1098000def9a53ba0a28d14f103b6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L800-L803
https://github.com/checkstyle/checkstyle/blob/e5c9f2c828c1098000def9a53ba0a28d14f103b6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L805-L810

This is already logging a DetailAST node. Nothing else needs to be done for the check, other than XPath suppression ITs.

## 8
https://github.com/checkstyle/checkstyle/blob/e5c9f2c828c1098000def9a53ba0a28d14f103b6/src/main/java/com/puppycrawl/tools/checkstyle/checks/javadoc/JavadocMethodCheck.java#L926-L932

This is the only opportunity to migrate to DetailAST logging. It uses `ExceptionInfo` which uses `ClassInfo` which uses `Token` which uses line and column numbers. However the `ClassInfo` is only really necessary to obtain the full identifier name via `FullIdent`. Instead of `ExceptionInfo` using the whole chain of dependencies we should be able to just store the full identifier and the corresponding AST node, as that is all we need.

I will wait on implementing changes until #7955 is merged. There will definitely be merge conflicts otherwise.

**Comment by romani** [comment]
@wltan , #7955 is on last review , please start doing this update.

**Comment by wltan** [comment]
> @wltan , #7955 is on last review , please start doing this update.

Ok. I already have the tests for the other three cases (2, 4, 5) done, will work on the last one (8) soon.

**Comment by timurt** [comment]
@wltan @romani 

I made a small research, how we can match `JavadocTag` object to corresponding `DetailAst` element. I noticed that `AbstractCheck.log(DetailAST ast, String key, Object... args)` and `AbstractCheck.log(int lineNo, int colNo, String key, Object... args)` use same logic for line and column number calculations, so I made conclusion that line and column in `JavadocTag` and corresponding `DetailAst` are the same.

I modified case **3** and **6** with the following implementation and it passed unit tests

```
                final DetailAstImpl detailAst = new DetailAstImpl();
                detailAst.setLineNo(tag.getLineNo());
                detailAst.setColumnNo(tag.getColumnNo());
                detailAst.setType(JavadocTokenTypes.JAVADOC_TAG);
                log(detailAst, MSG_UNUSED_TAG, "@param", arg1);
```

I am not sure it is the best solution, @romani what do you think? @wltan can test it with xpath regression test?

**Comment by wltan** [comment]
@timurt I tried the changes you suggested for case 3 and the expected violation output does match, but the Xpath query is completely empty which is not very useful.
```java
package org.checkstyle.suppressionxpathfilter.javadocmethod;

/**
 * Config:
 */
public class SuppressionXpathRegressionJavadocMethodSix {

    /**
     * @param x unused //warn
     */
    public void foo() {
    }

}
```
```java
    @Test
    public void testSix() throws Exception {
        final File fileToProcess =
                new File(getPath("SuppressionXpathRegressionJavadocMethodSix.java"));

        final DefaultConfiguration moduleConfig =
                createModuleConfig(JavadocMethodCheck.class);

        final String[] expectedViolation = {
            "9:8: " + getCheckMessage(JavadocMethodCheck.class, MSG_UNUSED_TAG,
                    "@param", "x"),
        };

        final List<String> expectedXpathQueries = Collections.singletonList("");

        runVerifications(moduleConfig, fileToProcess, expectedViolation,
                expectedXpathQueries);
    }
```
```
org.opentest4j.AssertionFailedError: Generated queries do not match expected ones ==> expected: java.util.Collections$SingletonList@730d2164<[]> but was: java.util.ArrayList@24959ca4<[]>
Expected :[]
Actual   :[]
```
To add on, I have no idea how to make the expected query match the actual in this case. I have tried `Collections.singletonList("")`, `Collections.emptyList()`, and `new ArrayList<>()` and none of them match.

**Comment by timurt** [comment]
@wltan yes, you are right, 
Currently, resulting detail ast of `JavaParser.parseFile(file, JavaParser.Options.WITH_COMMENTS)` does not contain Javadoc nodes, so it is impossible to access to these nodes using xpath

@romani 
We need to rethink about xpath query generator to parse javadoc comment blocks into `DetailAst` objects or add new option `JavaParser.Options.WITH_JAVADOC` to have resulting ast tree includes javadoc nodes

**Comment by romani** [comment]
Fix is merged
